File: cpp_files/raylib/shape_survivors/src/systems.cpp
#include "components.hpp"
#include "settings.hpp"
#include "systems.hpp"
#include <cmath>

namespace Systems {

// LEVEL + XP
int get_xp_next(const Components::Level& level) {
    return 10 + ((level.current * level.current) * 2);
}

// RENDERING
void draw_background_grid(void) {
    const float thickness  = 2.0f;
    const Color color      = LIGHTGRAY;
    const float block_size = 100.0f;

    for (float x = 0; x < MAP_SIZE; x += block_size) {
        DrawLineEx({x, 0}, {x, MAP_SIZE}, thickness, color);
    }

    for (float y = 0; y < MAP_SIZE; y += block_size) {
        DrawLineEx({0, y}, {MAP_SIZE, y}, thickness, color);
    }
}

void draw_rect(const Components::Position&  position,
               const Components::Rotation&  rotation,
               const Components::Rect&      rect,
               const Components::Color_&    color,
               const Components::Animation& animation) {
    const Rectangle dest   = {position.x, position.y, rect.width, rect.height};
    const Vector2   origin = {dest.width / 2.0f, dest.height / 2.0f};

    const Color tint = (animation.hit_timer > 0) ? RED : color.tint;

    DrawRectanglePro(dest, origin, rotation.angle, tint);
}

void draw_circle(const Components::Position& position,
                 const Components::Circle&   circle,
                 const Components::Color_&   color) {
    DrawCircleV({position.x, position.y}, circle.radius, color.tint);
}

// MOVEMENT
void set_direction(const Components::Position& position,
                   Components::Movement&       movement,
                   const Vector2               target_pos) {
    const float dx       = target_pos.x - position.x;
    const float dy       = target_pos.y - position.y;
    const float distance = std::sqrt((dx * dx) + (dy * dy));

    if (distance > 0) {
        movement.direction = {dx / distance, dy / distance};
    }
}

void move(Components::Position& position, const Components::Movement& movement, const float dt) {
    position.x += movement.direction.x * movement.speed * dt;
    position.y += movement.direction.y * movement.speed * dt;
}

void set_bounds(Components::Position& position,
                const float           half_w,
                const float           half_h,
                const Rectangle       bounds) {
    position.x = std::max(bounds.x + half_w, std::min(position.x, bounds.width - half_w));
    position.y = std::max(bounds.y + half_h, std::min(position.y, bounds.height - half_h));
}

// UPDATES
void update_xp(Components::Level& level, Components::Xp& xp) {
    if (xp.current >= xp.next) {
        xp.current -= xp.next;
        level.current += 1;
        level.pending += 1;
        xp.next = get_xp_next(level);
    }
}

void update_lifetime(Components::Lifetime& lifetime, const float dt) {
    lifetime.current -= dt;
    if (lifetime.current <= 0) {
        lifetime.current = 0;
    }
}

void update_hit_timer(Components::Animation& animation, const float dt) {
    animation.hit_timer -= dt;
    if (animation.hit_timer <= 0) {
        animation.hit_timer = 0;
    }
}

void update_health(Components::Health& health) {
    health.current = std::max(0, std::min(health.current, health.max));
}

} // namespace Systems

File: python_files/raylib_files/shape_survivors/systems.py
import math

import pyray as p

import components
import utils
from settings import MAP_SIZE


# LEVEL + XP
def get_xp_next(level: components.Level) -> int:
    return 10 + ((level.current**2) * 2)


def add_xp(level: components.Level, xp: components.Xp, amount: int) -> None:
    xp.current += amount

    while xp.current >= xp.next:
        xp.current -= xp.next
        level.current += 1
        level.pending += 1
        xp.next = get_xp_next(level)


# RENDERING
def draw_background_grid() -> None:
    thickness: float = 2.0
    color: p.Color = p.LIGHTGRAY
    block_size: int = 100

    for x in range(0, MAP_SIZE, block_size):
        p.draw_line_ex(p.Vector2(x, 0), p.Vector2(x, MAP_SIZE), thickness, color)

    for y in range(0, MAP_SIZE, block_size):
        p.draw_line_ex(p.Vector2(0, y), p.Vector2(MAP_SIZE, y), thickness, color)


def draw_rect(
    position: components.Position,
    rotation: components.Rotation,
    rect: components.Rect,
    color: components.Color,
    animation: components.Animation,
) -> None:
    dest: p.Rectangle = p.Rectangle(position.x, position.y, rect.width, rect.height)
    origin: p.Vector2 = p.Vector2(dest.width / 2.0, dest.height / 2.0)

    tint: p.Color = p.RED if animation.hit_timer > 0 else color.tint

    p.draw_rectangle_pro(dest, origin, rotation.angle, tint)


def draw_circle(
    position: components.Position, circle: components.Circle, color: components.Color
) -> None:
    p.draw_circle_v(p.Vector2(position.x, position.y), circle.radius, color.tint)


def draw_centered_text(
    position: components.Position, width: float, height: float, text: components.Text
) -> None:
    dest: p.Rectangle = p.Rectangle(position.x, position.y, width, height)

    utils.center_and_draw_text(
        utils.OriginType.CENTER,
        text.text,
        dest,
        text.font_size,
        text.spacing,
        text.tint,
    )


# MOVEMENT
def set_direction(
    position: components.Position, movement: components.Movement, target_pos: p.Vector2
) -> None:
    dx: float = target_pos.x - position.x
    dy: float = target_pos.y - position.y
    distance: float = math.sqrt(dx**2 + dy**2)
    if distance > 0:
        movement.direction = p.Vector2(dx / distance, dy / distance)


def move(
    position: components.Position, movement: components.Movement, dt: float
) -> None:
    position.x += movement.direction.x * movement.speed * dt
    position.y += movement.direction.y * movement.speed * dt


def set_bounds(
    position: components.Position, half_w: float, half_h: float, bounds: p.Rectangle
) -> None:
    position.x = max(bounds.x + half_w, min(position.x, bounds.width - half_w))
    position.y = max(bounds.y + half_h, min(position.y, bounds.height - half_h))


# UPDATES
def update_xp(level: components.Level, xp: components.Xp) -> None:
    if xp.current >= xp.next:
        xp.current -= xp.next
        level.current += 1
        level.pending += 1
        xp.next = get_xp_next(level)


def update_lifetime(lifetime: components.Lifetime, dt: float) -> None:
    lifetime.current -= dt
    if lifetime.current <= 0:
        lifetime.current = 0


def update_hit_timer(animation: components.Animation, dt: float) -> None:
    animation.hit_timer -= dt
    if animation.hit_timer <= 0:
        animation.hit_timer = 0


def update_health(health: components.Health) -> None:
    health.current = max(0, min(health.current, health.max))
